feat add windows compilation support

agrandville opened the pull request:
Let me know about your feedback on those changes.
Best,
Arnaud

agrandville committed:
feat add windows support

agrandville committed:
fix cmake minimal version

agrandville committed:
fix build set version

agrandville committed:
fix disable appveyor remote desktop

agrandville committed:
fix static cast size_t to CK_ULONG

error C2397: conversion from 'unsigned __int64' to 'CK_ULONG' requires a narrow
ing conversion

agrandville committed:
feat add Win64 compilation support

agrandville committed:
fix appveyor error 'version already exists'

agrandville committed:
fix windows CMake generator name

add vcpkg install gtest:x64-windows

agrandville committed:
fix downgrade to Visual Studio 14 2015

agrandville committed:
doc add windows compilation notes

agrandville commented:
@googlebot signed !

daviddrysdale commented on pkcs11test.cc:
```diff
@@ -83,22 +88,10 @@ void usage() {
   cerr << "  -v      : verbose output" << endl;
   cerr << "  -u pwd  : user PIN/password" << endl;
   cerr << "  -o pwd  : security officer PIN/password" << endl;
-  cerr << "  -w name : cipher to use for keys being wrapped, one of: { ";
```
It looks like code is getting lost along the way &ndash; please could you merge up to the latest codebase and re-generate the pull request.

agrandville replied:
merged

daviddrysdale commented on pkcs11test.cc:
```diff
@@ -121,13 +114,25 @@ CK_C_GetFunctionList load_pkcs11_library(const char* libpath, const char* libnam
-  void* lib = dlopen(fullname.c_str(), RTLD_NOW);
+  void* lib;
+#if defined WIN32
+  fullname = "C:\\Projects\\SoftHSMv2\\tmp32\\src\\lib\\Debug\\softhsm2.dll";
```
This looks specific to your particular machine/setup &ndash; please could we have something more generic/overridable.

daviddrysdale commented on third_party/win32/cryptoki.h:
```diff
@@ -0,0 +1,85 @@
+/* cryptoki.h include file for PKCS #11. */
```
The majority of this file is very similar to what's already in `pkcs11-env.h` &ndash; could you try modifying that file instead to have a few `#ifdef _WIN32` adaptations?

daviddrysdale commented on third_party/win32/getopt.cpp:
```diff
@@ -0,0 +1,519 @@
+/*-
```
Sorry to be a pain, but for importing external code please can you generate a separate commit that adds a pristine copy of the original files, with a commit message that gives thorough provenance, e.g something like:
```
Import pristine copy of NetBSD getopt code

Code taken from repository https://some/git/repo as of commit
1234567890abcdef ("summary line from commit message"), in
some/subdir/holding/getopt.{c,h}
```

Any local modifications should then happen in a separate commit.

Also, I think it probably makes more sense to use the subdir under `./third_party/` to indicate where the code came from (`./third_party/netbsd/` ?) rather than how it's used at the moment (`./third_party/win32/`).

agrandville committed:
feat publish github release

agrandville committed:
fix change artifact zip path

agrandville committed:
fix change artifact zip path

agrandville committed:
Update .appveyor.yml

agrandville committed:
Update CMakeLists.txt

daviddrysdale commented on digest.cc:
```diff
@@ -22,6 +22,13 @@
+#ifdef _WIN32
+#include <windows.h>
+#include <stdlib.h>
+#endif
```
Why is this needed?

daviddrysdale commented on pkcs11test.cc:
```diff
@@ -196,6 +193,10 @@ int main(int argc, char* argv[]) {
+//   CK_FUNCTION_LIST_PTR p11;
+//   (*get_fn_list)(&p11);
+//   p11->C_Initialize(NULL_PTR);
+
```
Needs tidying up?

daviddrysdale commented on CMakeLists.txt:
```diff
@@ -0,0 +1,82 @@
+#install(FILES $<TARGET_PDB_FILE:${PROJECT_NAME}> DESTINATION bin OPTIONAL)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/RelWithDebInfo/p11test.exe DESTINATION bin)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/RelWithDebInfo/p11test.pdb DESTINATION bin OPTIONAL)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/RelWithDebInfo/gmock.dll DESTINATION bin)
```
Is gmock.dll needed?

daviddrysdale commented on CMakeLists.txt:
```diff
@@ -0,0 +1,82 @@
+
+
+#install(TARGETS ${PROJECT_NAME} DESTINATION ${CMAKE_INSTALL_BINDIR})
+#install(FILES $<TARGET_PDB_FILE:${PROJECT_NAME}> DESTINATION bin OPTIONAL)
```
Please either fix or remove commented-out chunks.

daviddrysdale commented on .appveyor.yml:
```diff
@@ -0,0 +1,51 @@
+- cmd: ctest -C RelWithDebInfo --progress --verbose
+- cmd: cpack
+#on_finish:
+#  - ps: $blockRdp = $true; iex ((new-object net.webclient).DownloadString('https://raw.githubusercontent.com/appveyor/ci/master/scripts/enable-rdp.ps1'))
```
Please either fix or remove commented-out chunks.

daviddrysdale commented on .appveyor.yml:
```diff
@@ -0,0 +1,51 @@
+  - path: "*.zip"
+    name: artifacts-zip
+
+deploy:
```
Is a deploy section relevant?

agrandville committed:
Update .appveyor.yml

agrandville committed:
Update .appveyor.yml

agrandville committed:
Update .appveyor.yml

daviddrysdale reviewed (commented):
Thanks for the PR &ndash; adding Windows support sounds good.

The general approach seems reasonable, but on a quick first pass through the changes there's a few small details that would be good to get tidied up before merging.

agrandville committed:
feat release appVeyor builds

agrandville committed:
feat disable appveyor RDP on finish

agrandville committed:
fix PR regression

remove static path to p11 library

agrandville committed:
fix merge win32/cryptoki.h with pkcs11-env.h

fix move FILEPATH_NATIVE definition from pkcs11-env.h to pkcs11test.cc

agrandville committed:
fix remove getopt source

agrandville committed:
Import pristine copy of NetBSD getopt code

code taken from https://github.com/openbsd/src/blob/0025f5bb9c9be650004db9e38ab09eea44e2b392/lib/libc/stdlib/getopt_long.c and  https://github.com/openbsd/src/blob/0025f5bb9c9be650004db9e38ab09eea44e2b392/include/getopt.h

agrandville committed:
fix windows portability

agrandville committed:
fix update getopt build

agrandville committed:
fix remove unnecessary include

windows.h and stdlib.h

agrandville committed:
fix remove unnecessary GTest libraries

GTest::gtest_main
GTest::gmock
GTest::gmock_main

agrandville committed:
fix remove appveyor RDP on finish
